homeassistant/config_entries.py

```diff
@@ -297,6 +299,40 @@ def _validate_item(*, disabled_by: ConfigEntryDisabler | Any | None = None) -> N
         )
 
 
+class ConfigSubentryData(TypedDict):
+    """Container for configuration subentry data."""
+
+    data: Mapping[str, Any]
+    subentry_id: NotRequired[str]
```

Comment:
It's because config flows are allowed to provide subentries to be created when the config entry is created: https://github.com/home-assistant/core/pull/117355/files#diff-bcb932c3549daf52e1d6d8b02211ed0cc15557f5d17968f33a9daacf7ddac81bR2827, and the config flow implementation is not responsible for assigning the subentry id.

I agree it's unclear, maybe there should be two classes, one which is used by config flows when creating an entry to be pre-populated with subentries, which doesn't have the ID, and another, used only by config_entries.py, which always has the ID.

Reply:
Yeah, that would make it more clear.

Reply:
OK, changed in https://github.com/home-assistant/core/pull/117355/commits/219ef5d6219faf337bbac8d70264dab552d3b36e